plot-gui-lib/plot/private/gui/snip2d.rkt

```diff
@@ -243,9 +356,18 @@
       (super resize w h))
     ))
 
+(define plot-mouse-event-callback/c (-> (is-a?/c snip%) (is-a?/c mouse-event%)
+                                        (or/c real? #f) (or/c real? #f) any/c))
+
+(define/contract 2d-plot-snip+c%
```

Comment:
Using `define/contract` gives worse error messages than `contract-out` --- if a user gives a bad argument to a method, the error message will blame `plot-gui-lib/plot/private/gui/snip2d.rkt`

Here's a similar example, the outer module should get blamed, but the `define/contract` contract will blame the submodule `M`:  http://pasterack.org/pastes/35293

Suggestion:
- provide `2d-plot-ship%` and `make-2d-plot-snip` with `contract-out`
- then there's no need to define or document `2d-plot-snip+c%`, so remove it

(If you want `plot` to be able to make a `2d-plot-snip%` without a contract, then just move the `(contract-out [2d-plot-snip% ....])` to `plot-gui-lib/plot/snip.rkt`)

Reply:
I need some more guidance on your two suggestions.

>  *  provide 2d-plot-ship% and make-2d-plot-snip with contract-out

I'm not sure how to export `make-2d-plot-snip` with a contract.  First, it takes 7 arguments and I'm not sure what their contracts are, and `(-> any/c any/c ... 2d-plot-snip+c%)` would not be very elegant.  Second, `make-2d-plot-snip` creates a `2d-plot-snip%` instance and returns it.  Will `contract-out` attach the `2d-plot-snip+c%` class contract to an object?  I tried it out, see the diff at https://gist.github.com/alex-hhh/c8b9c9bb7de29a4c993cd99147fc9809, but this gives out an error when I try to create a plot:

```
instantiate: contract violation
  expected: class?
  given: 2d-plot-snip+c%
```

>  *  then there's no need to define or document 2d-plot-snip+c%, so remove it

We need to document the `set-mouse-event-callback` and `set-overlay-renderers` and the scribble documentation requires a class to wrap them (although my only experience with scribble is this pull request).   I could just use `snip%` , but that would be confusing in my opinion.  Or do you want to export `2d-plot-snip%` and use that symbol for the `@defclass`?

Reply:
The contract for `make-2d-plot-snip` should be `(unconstrained-domain-> (instanceof/c (class/c ....)))` with the `class/c` contract for `2d-plot-snip%`.

(I just remembered `unconstrained-domain->` this morning. If that didn't exist, then I'd suggest the 7 `any/c`s --- because the point of the contract is to wrap the objects this function returns, not to check its arguments [which all come from the plot library])

For documentation, I just want to export `2d-plot-snip%`.